File: backend/token_saver.py
import re

class TokenSaver:
    def __init__(self, max_context_messages=5):
        self.max_context_messages = max_context_messages
        self.filler_words = ["uh", "um", "er", "ah", "like", "you know", "basically", "actually"]

    def clean_text(self, text):
        # Remove common filler words (case insensitive)
        for word in self.filler_words:
            pattern = re.compile(rf'\s*\b{word}\b\s*', re.IGNORECASE)
            text = pattern.sub(' ', text)

        # Remove extra spaces
        text = " ".join(text.split())

        # Remove redundant punctuation
        text = re.sub(r'([!?.,]){2,}', r'\1', text)

        # Remove leading/trailing punctuation that might be left after filler word removal
        text = re.sub(r'^\W+', '', text)

        return text.strip()

    def trim_context(self, messages):
        """Keep only last N messages"""
        if len(messages) > self.max_context_messages:
            return messages[-self.max_context_messages:]
        return messages

    def optimize_prompt(self, messages):
        """Convert list of messages into a structured short prompt"""
        prompt = ""
        for msg in messages:
            role = msg.get("role", "user")
            content = self.clean_text(msg.get("content", ""))
            if role == "user":
                prompt += f"User: {content}\n"
            else:
                prompt += f"AI: {content}\n"
        prompt += "AI:"
        return prompt

File: backend_csharp/Services/TokenSaver.cs
using System.Text.RegularExpressions;

namespace LightAI.Backend.Services;

public class TokenSaver
{
    private readonly int _maxContextMessages;
    private readonly string[] _fillerWords = { "uh", "um", "er", "ah", "like", "you know", "basically", "actually" };

    public TokenSaver(int maxContextMessages = 5)
    {
        _maxContextMessages = maxContextMessages;
    }

    public string CleanText(string text)
    {
        if (string.IsNullOrWhiteSpace(text)) return string.Empty;

        // Remove filler words
        foreach (var word in _fillerWords)
        {
            text = Regex.Replace(text, $@"\s*\b{word}\b\s*", " ", RegexOptions.IgnoreCase);
        }

        // Normalize whitespace (preserving single newlines)
        text = Regex.Replace(text, @"[^\S\r\n]+", " ");
        text = Regex.Replace(text, @"\n{2,}", "\n");

        // Normalize redundant punctuation
        text = Regex.Replace(text, @"([!?.,]){2,}", "$1");

        // Remove leading non-word characters except ! and /
        text = Regex.Replace(text, @"^[^\w!/]+", "");

        return text.Trim();
    }

    public List<T> TrimContext<T>(List<T> messages)
    {
        if (messages.Count > _maxContextMessages)
        {
            return messages.Skip(messages.Count - _maxContextMessages).ToList();
        }
        return messages;
    }

    public string OptimizePrompt(List<ChatMessage> messages)
    {
        var prompt = "";
        foreach (var msg in messages)
        {
            var content = CleanText(msg.Content);
            var role = msg.Role.ToLower() == "user" ? "User" : "AI";
            prompt += $"{role}: {content}\n";
        }
        prompt += "AI:";
        return prompt;
    }
}

public class ChatMessage
{
    public string Role { get; set; } = string.Empty;
    public string Content { get; set; } = string.Empty;
}
